File: app/src/main/java/com/bonustrack02/tp08goodprice/JapaneseFragment.java
package com.bonustrack02.tp08goodprice;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;

public class JapaneseFragment extends Fragment {

    ArrayList<Item> items = new ArrayList<>();
    RecyclerView recyclerView;
    RecyclerAdapter adapter;
    ProgressBar progressBar;
    String apiKey = "4e4c586954776c7338365056587546";

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_japanese, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        progressBar = view.findViewById(R.id.progressbar);
        adapter = new RecyclerAdapter(getActivity(), items);
        recyclerView = view.findViewById(R.id.recycler);

        loadApiData();

        recyclerView.setAdapter(adapter);
    }

    void loadApiData() {
        Thread thread = new Thread() {

            @Override
            public void run() {
                int startIndex = 1, endIndex = 30;

                String apiAddress = "http://openapi.seoul.go.kr:8088/"
                        + apiKey + "/xml/ListPriceModelStoreService/"
                        + startIndex+ "/" + endIndex + "/003/";
                try {
                    URL url = new URL(apiAddress);

                    InputStream inputStream = url.openStream();
                    InputStreamReader reader = new InputStreamReader(inputStream);

                    XmlPullParserFactory factory = XmlPullParserFactory.newInstance();
                    XmlPullParser parser = factory.newPullParser();
                    parser.setInput(reader);

                    int eventType = parser.getEventType();

                    Item item = null;

                    while (eventType != XmlPullParser.END_DOCUMENT) {
                        switch (eventType) {
                            case XmlPullParser.START_DOCUMENT:
                                getActivity().runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        progressBar.setVisibility(View.VISIBLE);
                                    }
                                });
                                break;

                            case XmlPullParser.START_TAG:
                                String startTag = parser.getName();
                                if (startTag.equals("row")) {
                                    item = new Item();
                                    Log.i("Tag", "row");
                                } else if (startTag.equals("SH_NAME")) {
                                    parser.next();
                                    if (item != null) item.name = parser.getText();
                                } else if (startTag.equals("SH_ADDR")) {
                                    parser.next();
                                    if (item != null) item.address = parser.getText();
                                } else if (startTag.equals("SH_PHONE")) {
                                    parser.next();
                                    if (item != null) item.phone = parser.getText();
                                } else if (startTag.equals("SH_PHOTO")) {
                                    parser.next();
                                    if (item != null) item.imgShop = parser.getText();
                                }
                                break;

                            case XmlPullParser.END_TAG:
                                String endTag = parser.getName();
                                if (endTag.equals("row")) if (item != null) items.add(item);
                                break;

                            case XmlPullParser.TEXT:
                                break;
                        }
                        eventType = parser.next();
                    }

                    getActivity().runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            if (adapter != null) adapter.notifyDataSetChanged();
                            Log.i("Tag", "" + items.size());
                            progressBar.setVisibility(View.GONE);
                        }
                    });

                } catch (MalformedURLException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                } catch (XmlPullParserException e) {
                    e.printStackTrace();
                }
            }
        };
        thread.start();
    }

}

File: app/src/main/java/com/bonustrack02/tp08goodprice/JapaneseFragment.kt
package com.bonustrack02.tp08goodprice

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bonustrack02.tp08goodprice.databinding.FragmentJapaneseBinding
import com.google.android.material.snackbar.Snackbar
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class JapaneseFragment: Fragment() {
    private val binding: FragmentJapaneseBinding by lazy { FragmentJapaneseBinding.inflate(layoutInflater) }
    private val items = mutableListOf<Shop>()
    private val apiKey = BuildConfig.APIKEYSEOUL
    private var startIndex = 1
    private var endIndex = 30
    private var totalCount = 0


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.recycler.adapter = RecyclerAdapter(requireContext(), items)
        getStoreListUsingRetrofit(startIndex, endIndex)
        binding.progressbar.visibility = View.VISIBLE

        binding.recycler.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)

                val lastItemPosition = (recyclerView.layoutManager as LinearLayoutManager).findLastCompletelyVisibleItemPosition()
                val totalItemCount = recyclerView.adapter?.itemCount?.minus(1)

                if (lastItemPosition == totalItemCount) {
                    this@JapaneseFragment.startIndex += 30
                    this@JapaneseFragment.endIndex += 30
                    getStoreListUsingRetrofit(this@JapaneseFragment.startIndex, this@JapaneseFragment.endIndex)

                    if (totalCount == items.size) Snackbar.make(context!!, binding.snackbarContainer, "마지막 리스트입니다.", Snackbar.LENGTH_SHORT).show()
                }
            }
        })
    }

    private fun getStoreListUsingRetrofit(startIndex: Int, endIndex: Int) {
        val retrofitService = RetrofitHelper.getInstance().create(RetrofitService::class.java)
        val call = retrofitService.getStoreJson(apiKey, startIndex, endIndex, "003")
        call.enqueue(object : Callback<RetrofitResponse> {
            override fun onResponse(
                call: Call<RetrofitResponse>,
                response: Response<RetrofitResponse>
            ) {
                Log.d("retrofit warning", "${response.body()}")
                Log.d("retrofit result", "${response.raw()}")
                response.body()!!.responseItem.ShopList.forEach {
                    if (items.contains(it)) return@forEach
                    items.add(it)
                }
                binding.recycler.adapter?.notifyDataSetChanged()
                Log.d("recycler size", "${items.size}")
                totalCount = response.body()!!.responseItem.totalCount
                binding.progressbar.visibility = View.GONE
            }

            override fun onFailure(
                call: Call<RetrofitResponse>,
                t: Throwable
            ) {
                Log.e("retrofit error", "${t.stackTrace}")
            }
        })
    }
}